src/crawlee/storage_clients/_sql/_dataset_client.py

```diff
@@ -0,0 +1,305 @@
+
+    _METADATA_TABLE = DatasetMetadataDB
+    """SQLAlchemy model for dataset metadata."""
+
+    _ITEM_TABLE = DatasetItemDB
+    """SQLAlchemy model for dataset items."""
+
+    _CLIENT_TYPE = 'Dataset'
+    """Human-readable client type for error messages."""
+
+    def __init__(
+        self,
+        *,
+        id: str,
+        storage_client: SQLStorageClient,
+    ) -> None:
+        """Initialize a new instance.
+
+        Preferably use the `SqlDatasetClient.open` class method to create a new instance.
+        """
+        super().__init__(id=id, storage_client=storage_client)
+
+    @override
+    async def get_metadata(self) -> DatasetMetadata:
+        """Get dataset metadata from the database."""
+        # The database is a single place of truth
+        metadata = await self._get_metadata(DatasetMetadata)
+        return cast('DatasetMetadata', metadata)
+
+    @classmethod
+    async def open(
+        cls,
+        *,
+        id: str | None,
+        name: str | None,
+        storage_client: SQLStorageClient,
+    ) -> SQLDatasetClient:
+        """Open an existing dataset or create a new one.
+
+        Args:
+            id: The ID of the dataset to open. If provided, searches for existing dataset by ID.
+            name: The name of the dataset to open. If not provided, uses the default dataset.
+            storage_client: The SQL storage client instance.
+
+        Returns:
+            An instance for the opened or created storage client.
+
+        Raises:
+            ValueError: If a dataset with the specified ID is not found.
+        """
+        return await cls._safely_open(
+            id=id,
+            name=name,
+            storage_client=storage_client,
+            metadata_model=DatasetMetadata,
+            extra_metadata_fields={'itemCount': 0},
+        )
+
+    @override
+    async def drop(self) -> None:
+        """Delete this dataset and all its items from the database.
+
+        This operation is irreversible. Uses CASCADE deletion to remove all related items.
+        """
+        await self._drop()
+
+    @override
+    async def purge(self) -> None:
+        """Remove all items from this dataset while keeping the dataset structure.
+
+        Resets item_count to 0 and deletes all records from dataset_item table.
+        """
+        await self._purge(
+            metadata_kwargs={'new_item_count': 0, 'update_accessed_at': True, 'update_modified_at': True, 'force': True}
+        )
+
+    @override
+    async def push_data(self, data: list[dict[str, Any]] | dict[str, Any]) -> None:
+        """Add new items to the dataset."""
+        if not isinstance(data, list):
+            data = [data]
+
+        db_items: list[dict[str, Any]] = []
+
+        for item in data:
+            # Serialize with default=str to handle non-serializable types like datetime
+            json_item = json.dumps(item, default=str, ensure_ascii=False)
```

Comment:
I'd consider using [JSONType](https://github.com/kvesteri/sqlalchemy-utils/blob/master/sqlalchemy_utils/types/json.py) from sqlalchemy-utils. Using the built-in JSON in postgres would unlock some cool aggregation options for people who want to interact with the DB directly (I think that's a legit use case).